File: menu_v2.py
import sys
import struct
import termios

def func1(x, y):
    print "%d+%d = %d" %(x, y, x+y)
def func2(x, y):
    print "%d-%d = %d" %(x, y, x-y)
def func3(x, y):
    print "%d*%d = %d" %(x, y, x*y)

funcdict = {
        "func1": func1,
        "func2": func2,
        "func3": func3,
        }

menu = [
        #(index, id, function name, function ptr)
        ("0", '0', "Exit", "NULL"),
        ("1", '1', "Func#1", "func1"),
        ("a", 'a', "Func#2", "func2"),
        ("^U", '\x15', "Func#3", "func3"),
       ]

init_setting = termios.tcgetattr(sys.stdin.fileno())
new_setting = termios.tcgetattr(sys.stdin.fileno())

new_setting[3] &= ~termios.ICANON
new_setting[3] &= ~termios.ECHO
new_setting[6][termios.VTIME] = 0
new_setting[6][termios.VMIN] = 1

while True:
    for i in range(len(menu)):
        print "[%s]:\t%s." %(menu[i][0], menu[i][2])

    termios.tcsetattr(sys.stdin.fileno(), termios.TCSAFLUSH, new_setting)
    opt = sys.stdin.read(1)
    termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, init_setting)

    for i in range(len(menu)):
        if opt == menu[i][1]:
            if menu[i][3] == "NULL":
                sys.exit()
            else:
                funcdict[menu[i][3]](3, 2)

File: menu_v2.c
#include <stdlib.h>
#include <stdio.h>
#include <termios.h>	//termios struct, tcgetattr(), tcsetattr()
#include <string.h>

#define ARRAY_SIZE(x)  (sizeof(x) / sizeof((x)[0]))

int func1(int x, int y)
{
    printf("%d+%d = %d\n", x, y, x+y);

    return 0;
}

int func2(int x, int y)
{
    printf("%d-%d = %d\n", x, y, x-y);

    return 0;
}

int func3(int x, int y)
{
    printf("%d*%d = %d\n", x, y, x*y);

    return 0;
}

struct option_attr{
    char* id;
    char opt;
    char *title;
    int (*func)(int arg1, int arg2);
};

struct option_attr menu[] = {
    {
        .id = "0",
        .opt = '0',
        .title = "Exit.",
        .func = NULL,
    },
    {
        .id = "1",
        .opt = '1',
        .title = "function #1",
        .func = &func1,
    },
    {
        .id = "a",
        .opt = 'a',
        .title = "function #2",
        .func = &func2,
    },
    {
        .id = "^U",
        .opt = 0x15,
        .title = "function #3",
        .func = &func3,
    },
};

int main()
{
	char tmp[1];
	struct termios init_setting, new_setting;
	unsigned int op, i;

	tcgetattr(fileno(stdin), &init_setting);

	new_setting = init_setting;
	new_setting.c_lflag &= ~ICANON;
	new_setting.c_lflag &= ~ECHO;
	new_setting.c_cc[VTIME]=0;
	new_setting.c_cc[VMIN]=1;

	while(1){

        for(i = 0; i < ARRAY_SIZE(menu); i++){
            printf("[%s]: %s\n", menu[i].id, menu[i].title);
        }

        tcsetattr(fileno(stdin), TCSAFLUSH, &new_setting);

        do{
            *tmp = getchar();
        }while(*tmp=='\n' || *tmp=='\r');

        tcsetattr(fileno(stdin), TCSANOW, &init_setting);

        for(i = 0; i < ARRAY_SIZE(menu); i++){
            if(*tmp == menu[i].opt){
                if(menu[i].func == NULL)
                    return 0;
                else
                    menu[i].func(3, 2);
            }
        }
    }
	return 0;
}
